### PR TITLE
Feat: revisit fork handling

### PR DESCRIPTION
**Pull request by lgalabru**
In this PR, we are improving the Oreo's ability to relay a consistent view of the chain, by supporting forks and unexpected network behaviors.
We were able to live without it in Clarinet because `clarinet integrate` spins up local, solo-mining networks - not supposed to be adversarial.
Thanks to this effort, chainhooks can now reasonably be used on other networks like Testnet and Mainnet. 
I'm planning on opening a subsequent PR with an additional batch of 50 vectors, probably end of next week.

### CODE REVIEW

File: vendor/orchestra-event-observer/src/indexer/bitcoin/blocks_pool.rs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+        // Select canonical fork
+        let mut canonical_fork_id = 0;
+        let mut highest_height = 0;
+        let mut highest_btc_spent = 0;
+        for (fork_id, fork) in self.forks.iter() {
+            info!("Active fork: {} - {}", fork_id, fork);
+            if fork.get_length() >= highest_height {
+                highest_height = fork.get_length();
+                if fork.amount_of_btc_spent > highest_btc_spent
```
Comment (obycode) [inline vendor/orchestra-event-observer/src/indexer/bitcoin/blocks_pool.rs:134]: What is the meaning of `amount_of_btc_spent` for a bitcoin fork?

File: vendor/orchestra-event-observer/src/indexer/stacks/blocks_pool.rs [hunk 2]
```diff
@@ -0,0 +1,902 @@
...
+
+        // Select canonical fork
+        let mut canonical_fork_id = 0;
+        let mut highest_height = 0;
+        let mut highest_btc_spent = 0;
+        for (fork_id, fork) in self.forks.iter() {
+            info!("Active fork: {} - {}", fork_id, fork);
+            if fork.get_length() >= highest_height {
+                highest_height = fork.get_length();
+                if fork.amount_of_btc_spent > highest_btc_spent
```
Comment (obycode) [inline vendor/orchestra-event-observer/src/indexer/stacks/blocks_pool.rs:144]: Where does `amount_of_btc_spent` get set?

File: vendor/orchestra-event-observer/src/indexer/stacks/blocks_pool.rs [hunk 3]
```diff
@@ -0,0 +1,902 @@
...
+    // condider the 3 following forks
+    //
+    // 1) A1 - B1 - a1 - b1 - c1 - C1
+    //
+    // 2) A1 - B1 - a1 - b1 - C2
+    //
+    // 3) A1 - B1 - a1 - b1 - c1 - d1 - C3
+    //
+    // How can we always be sending back the right diff?
+    // As is, if 1) 2) #), we will be sending:
```
Comment (obycode) [inline vendor/orchestra-event-observer/src/indexer/stacks/blocks_pool.rs:536]:
```suggestion
    // As is, if 1) 2) 3), we will be sending:
```

**Review by obycode — COMMENTED** [review]
This is looking good! Mostly minor comments.

File: vendor/orchestra-event-observer/src/indexer/stacks/blocks_pool.rs (hunk 2 above)
Comment (lgalabru) [reply vendor/orchestra-event-observer/src/indexer/stacks/blocks_pool.rs:144]: Good catch, I'll handle this in the next iteration - https://github.com/hirosystems/orchestra/issues/5

File: vendor/orchestra-event-observer/src/indexer/bitcoin/blocks_pool.rs (hunk 1 above)
Comment (lgalabru) [reply vendor/orchestra-event-observer/src/indexer/bitcoin/blocks_pool.rs:134]: copy pasta!

File: vendor/orchestra-types-js/src/index.ts [hunk 4]
```diff
@@ -1,5 +1,65 @@
...
+ * @interface StacksChainUpdate
+ */
+export interface StacksChainUpdate {
+  /**
+   * @type {Array<StacksBlockUpdate>}
+   * @memberof StacksChainUpdate
+   */
+  new_blocks: Array<StacksBlockUpdate>;
+  /**
+   * @type {Array<StacksBlockUpdate>}
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-types-js/src/index.ts:31]: This is actually a `StacksBlockUpdate` - it hold new blocks alongs with other data fields (parent micro blocks to rollback / update)

File: node-bindings/src/lib.rs [hunk 5]
```diff
@@ -115,15 +116,15 @@ impl StacksDevnet {
                 while let Ok(ref event) = devnet_rx.recv() {
                     match event {
                         DevnetEvent::BitcoinChainEvent(
-                            BitcoinChainEvent::ChainUpdatedWithBlock(block),
+                            BitcoinChainEvent::ChainUpdatedWithBlocks(block),
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/src/lib.rs:119]: Addressed with latest

File: node-bindings/src/lib.rs [hunk 6]
```diff
@@ -115,15 +116,15 @@ impl StacksDevnet {
...
                         ) => {
                             bitcoin_block_tx
                                 .send(block.clone())
                                 .expect("Unable to transmit bitcoin block");
                         }
-                        DevnetEvent::StacksChainEvent(StacksChainEvent::ChainUpdatedWithBlock(
-                            block,
-                        )) => {
+                        DevnetEvent::StacksChainEvent(
+                            StacksChainEvent::ChainUpdatedWithBlocks(block),
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/src/lib.rs:126]: Addressed with latest

File: src/integrate/chains_coordinator.rs [hunk 7]
```diff
@@ -210,7 +210,7 @@ pub async fn start_chains_coordinator(
                 // Contextual shortcut: Devnet is an environment under control,
                 // with 1 miner. As such we will ignore Reorgs handling.
                 let (log, status) = match &chain_update {
-                    BitcoinChainEvent::ChainUpdatedWithBlock(block) => {
+                    BitcoinChainEvent::ChainUpdatedWithBlocks(block) => {
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/integrate/chains_coordinator.rs]: Addressed with latest

File: vendor/orchestra-event-observer/src/chainhooks/mod.rs [hunk 8]
```diff
@@ -11,273 +13,399 @@ use bitcoincore_rpc::bitcoin::{Address, PubkeyHash, PublicKey, Script};
...
+                                }
+                            }
+                            _ => {}
                         }
                     }
                 }
+                (
+                    StacksTransactionKind::NativeTokenTransfer,
+                    StacksHookPredicate::StxEvent(_expected_stx_event),
+                ) => {}
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-event-observer/src/chainhooks/mod.rs:304]: Addressed with latest

File: vendor/orchestra-event-observer/src/chainhooks/mod.rs [hunk 9]
```diff
@@ -11,273 +13,399 @@ use bitcoincore_rpc::bitcoin::{Address, PubkeyHash, PublicKey, Script};
...
+            }
+            if let StacksTransactionKind::ContractCall(_actual_contract_call) = &tx.metadata.kind {
+                match &chainhook.predicate {
+                    StacksHookPredicate::ContractCall(_expected_contract_call) => {}
+                    StacksHookPredicate::PrintEvent(_expected_print_event) => {}
+                    StacksHookPredicate::StxEvent(_expected_stx_event) => {}
+                    StacksHookPredicate::NftEvent(_expected_nft_event) => {}
+                    StacksHookPredicate::FtEvent(_expected_ft_event) => {}
+                }
             }
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-event-observer/src/chainhooks/mod.rs:315]: Addressed with latest

File: vendor/orchestra-event-observer/src/chainhooks/mod.rs [hunk 10]
```diff
@@ -11,273 +13,399 @@ use bitcoincore_rpc::bitcoin::{Address, PubkeyHash, PublicKey, Script};
...
+
+                for block in event.new_blocks.iter() {
+                    for tx in block.transactions.iter() {
+                        if chainhook.evaluate_predicate(&tx) {
+                            apply.push((tx, &block.block_identifier))
+                        }
                     }
                 }
+
+                if !apply.is_empty() || !rollback.is_empty() {
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-event-observer/src/chainhooks/mod.rs:340]: Addressed with latest

File: vendor/orchestra-event-observer/src/chainhooks/mod.rs [hunk 11]
```diff
@@ -11,273 +13,399 @@ use bitcoincore_rpc::bitcoin::{Address, PubkeyHash, PublicKey, Script};
...
-                    "transaction": tx,
-                    "proof": proof,
-                    "block_identifier": block_identifier,
-                    "confirmations": 1,
-                })],
+                "apply": trigger.apply.into_iter().map(|(transaction, block_identifier)| {
+                    json!({
+                        "transaction": transaction,
+                        "block_identifier": block_identifier,
+                        "confirmations": 1,
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-event-observer/src/chainhooks/mod.rs:395]: Addressed with latest

File: vendor/orchestra-types-js/src/index.ts [hunk 12]
```diff
@@ -1,5 +1,65 @@
+/**
+ * BitcoinChainUpdate provide informations about new blocks and confirmed blocks.
+ * @export
+ * @interface BitcoinChainUpdate
+ */
+export interface BitcoinChainUpdate {
+  /**
+   * @type {Array<Block>}
+   * @memberof ChainUpdate
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-types-js/src/index.ts:9]: Addressed with latest

File: vendor/orchestra-types-js/src/index.ts [hunk 13]
```diff
@@ -1,5 +1,65 @@
...
+   */
+  new_blocks: Array<Block>;
+  /**
+   * @type {Array<Block>}
+   * @memberof ChainUpdate
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-types-js/src/index.ts:14]: Addressed with latest

File: vendor/orchestra-types-js/src/index.ts [hunk 14]
```diff
@@ -8,6 +68,7 @@ export interface BitcoinChainEvent {
 
 export interface StacksChainEvent {
   apply: StacksApply[];
+  rollback: StacksRollback;
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-types-js/src/index.ts:71]: Addressed with latest

File: vendor/orchestra-event-observer/src/indexer/stacks/blocks_pool.rs (hunk 3 above)
Comment (lgalabru) [reply vendor/orchestra-event-observer/src/indexer/stacks/blocks_pool.rs:536]: Addressed with latest

File: vendor/orchestra-types-js/src/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (obycode) [reply vendor/orchestra-types-js/src/index.ts:31]: Should the type match then?

```rust
  confirmed_blocks: Array<Block>;
```

**Review by obycode — DISMISSED** [review]
LGTM!

File: vendor/orchestra-types-js/src/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lgalabru) [reply vendor/orchestra-types-js/src/index.ts]: Addressed with latest

**Pull request merged**